src/connectors/query-rules/connectQueryRules.ts

```diff
@@ -0,0 +1,237 @@
+  [facetName: string]: (facetValues: string[]) => string[];
+};
+type ParamTransformRuleContexts = (ruleContexts: string[]) => string[];
+type ParamTransformItems = (items: object[]) => any;
+
+export type QueryRulesConnectorParams = {
+  trackedFilters?: ParamTrackedFilters;
+  transformRuleContexts?: ParamTransformRuleContexts;
+  transformItems?: ParamTransformItems;
+};
+
+export interface QueryRulesRenderOptions<T> extends RenderOptions<T> {
+  userData: object[];
+}
+
+export type QueryRulesRenderer<T> = Renderer<QueryRulesRenderOptions<T>>;
+
+export type QueryRulesWidgetFactory<T> = WidgetFactory<
+  QueryRulesConnectorParams & T
+>;
+
+export type QueryRulesConnector = {
+  <T>(
+    render: QueryRulesRenderer<T>,
+    unmount?: () => void
+  ): QueryRulesWidgetFactory<T>;
+};
+
+// A context rule must consist only of alphanumeric characters, hyphens, and underscores.
+// See https://www.algolia.com/doc/guides/managing-results/refine-results/merchandising-and-promoting/in-depth/implementing-query-rules/#context
+function escapeRuleContext(ruleName: string) {
+  return ruleName.replace(/[^a-z0-9-_]+/gi, '_');
+}
+
+function getRuleContextsFromTrackedFilters({
+  helper,
+  trackedFilters,
+}: {
+  helper: Helper;
+  trackedFilters: ParamTrackedFilters;
+}) {
+  const ruleContexts = Object.keys(trackedFilters).reduce(
+    (facets: string[], facetName) => {
+      const getFacetValues = trackedFilters[facetName];
+      const facetRefinements: string[] = helper
+        .getRefinements(facetName)
+        .map((refinement: FacetRefinement) => refinement.value);
+      const facetValues = getFacetValues(facetRefinements);
+
+      return [
+        ...facets,
+        ...facetRefinements
+          .filter(
+            facetRefinement =>
+              facetValues.findIndex(
+                facetValue => facetValue === facetRefinement
+              ) !== -1
+          )
+          .map(facetValue =>
+            escapeRuleContext(`ais-${facetName}-${facetValue}`)
+          ),
+      ];
+    },
+    []
+  );
+
+  return ruleContexts;
+}
+
+function prepareRuleContexts({
+  initialRuleContexts,
+  newRuleContexts,
+  transformRuleContexts,
+}: {
+  initialRuleContexts: string[];
+  newRuleContexts: string[];
+  transformRuleContexts: ParamTransformRuleContexts;
+}) {
+  const allRuleContexts = [...initialRuleContexts, ...newRuleContexts];
+
+  warning(
+    allRuleContexts.length <= 10,
+    `
+The maximum number of \`ruleContexts\` is 10. They have been sliced to that limit.
+Consider using \`transformRuleContexts\` to minimize the number of rules sent to Algolia.
+`
+  );
+
+  const ruleContexts = transformRuleContexts(allRuleContexts).slice(0, 10);
+
+  return ruleContexts;
+}
+
+function applyRuleContexts({ helper, ruleContexts }) {
+  const previousRuleContexts = helper.getQueryParameter('ruleContexts');
+
+  if (!isEqual(previousRuleContexts, ruleContexts)) {
+    helper.setQueryParameter('ruleContexts', ruleContexts).search();
```

Comment:
When refining an item in a refinement list for example, this will trigger 2 network requests because of this implementation.

1. Send request to refine the filter
2. Send request with the new `ruleContexts`

I'm not sure how to avoid that.

Reply:
Right now there is none. It's fine for now, we can add the batch to the list of things that the helper can do for us.

Reply:
Actually after talking a bit with Marie & Julien it might be problematic. We can have a blink:

1. select the refinement -> a request is triggered
2. render with the results, context is applied -> a new request is triggered
3. a query rule is triggered based on the context that set a new filter
4. render with other results, which means that at the step 2 we get stalled results

Reply:
Is that an acceptable tradeoff for now? (I would say so)

Reply:
It worth trying with a query rule that alter the results to see the actual impact. We can decide then if it's a tradeoff that is acceptable or not.

Reply:
I promoted the movie "Gone girl" to position 2 in [this movie demo](https://movies-query-rules.now.sh/) when you click on the genre "Drama".

There's indeed a flash, which is noticeable.